cheddar/cheddar-application/src/test/java/com/clicktravel/cheddar/application/retry/RetryableAspectTest.java

```diff
@@ -0,0 +1,250 @@
+        try {
+            retryableAspect.attemptMethodAndRetryIfNeeded(mockProceedingJoinPoint, mockRetryable);
+        } catch (final RetryAspectTestException e) {
+            actualException = e;
+        }
+
+        // Then
+        verify(mockProceedingJoinPoint, times(retryMaxAttempts)).proceed();
+
+        assertNotNull(actualException);
+    }
+
+    @Test
+    public void shouldAttemptMethodAndFailImmedaitelyOnException_withImmediateFailureExceptionClassThrownAndRetryEnabled()
+            throws Throwable {
+        // Given
+        final RetryableAspect retryableAspect = new RetryableAspect();
+        final ProceedingJoinPoint mockProceedingJoinPoint = setupSimpleProceedingJoinPointMock();
+        final int retryMaxAttempts = 3;
+        final Retryable mockRetryable = setupMockRetryable(retryMaxAttempts, 2000);
+
+        RetryableConfiguration.setRetryableEnabled(true);
+        when(mockRetryable.failImmediatelyOn()).thenReturn(new Class[] { RetryAspectTestException.class });
+        when(mockProceedingJoinPoint.proceed()).thenThrow(new RetryAspectTestException());
+
+        // When
+        RetryAspectTestException actualException = null;
+        try {
+            retryableAspect.attemptMethodAndRetryIfNeeded(mockProceedingJoinPoint, mockRetryable);
+        } catch (final RetryAspectTestException e) {
+            actualException = e;
+        }
+
+        // Then
+        verify(mockProceedingJoinPoint, atMost(1)).proceed();
+
+        assertNotNull(actualException);
+    }
+
+    @Test
+    public void shouldAttemptMethodAndRetry_withExceptionThrownAndRetryEnabledAndDelayAdjusted() throws Throwable {
+        // Given
+        final RetryableAspect retryableAspect = new RetryableAspect();
+        final ProceedingJoinPoint mockProceedingJoinPoint = setupSimpleProceedingJoinPointMock();
+        final int retryDelayInMillis = 3000;
+        final int maxRetries = randomIntInRange(2, 6);
+        final int expectedRuntime = retryDelayInMillis * (maxRetries - 1);
+        final Retryable mockRetryable = setupMockRetryable(maxRetries, retryDelayInMillis);
+
+        RetryableConfiguration.setRetryableEnabled(true);
+        when(mockProceedingJoinPoint.proceed()).thenThrow(new RetryAspectTestException());
+
+        // When
+        DateTime startDateTime = null;
+        DateTime endDateTime = null;
+        RetryAspectTestException actualException = null;
+        try {
+            startDateTime = DateTime.now();
+            retryableAspect.attemptMethodAndRetryIfNeeded(mockProceedingJoinPoint, mockRetryable);
+        } catch (final RetryAspectTestException e) {
+            endDateTime = DateTime.now();
+            actualException = e;
+        }
+
+        // Then
+        verify(mockProceedingJoinPoint, times(maxRetries)).proceed();
+
+        assertNotNull(actualException);
+        assertTrue(startDateTime.plusMillis(expectedRuntime).getMillis() <= endDateTime.getMillis());
```

Comment:
I don't think this is a robust test, the actual elapsed time may vary from the expected time either way. Rather than have a slow running test which may fail, a more robust alternative is to have the following:
```java
const long EPSILON_MS = 300; // test timing tolerance
int retryDelayInMillis = 400;
int maxAttempts = 3; // 2 retries * 400ms delay = 800ms expected runtime
...
long elapsedMillis = DateTime.now().getMillis - startDateTime;
assertTrue(Math.abs(elapsedMillis - expectedRuntime) < EPSILON_MS);
```